paddle/fluid/operators/batch_norm_op.cu

```diff
@@ -234,6 +234,50 @@ static __global__ void KeBNBackwardData(const T *dy,
   }
 }
 
+template <typename X, typename Scale, typename Bias, typename Mean,
+          typename Variance, typename Y>
+static __global__ void KeBNRestoreData(const framework::DataLayout layout, X *x,
+                                       const Scale *scale, const Bias *bias,
+                                       const Mean *mean,
+                                       const Variance *variance, double epsilon,
+                                       int C, int M, const int num,
+                                       const Y *y) {
+  int gid = blockIdx.x * blockDim.x + threadIdx.x;
+  int stride = blockDim.x * gridDim.x;
+  for (int i = gid; i < num; i += stride) {
+    const int c = layout == framework::DataLayout::kNCHW ? (i / M) % C : i % C;
+    x[i] = (y[i] - bias[c]) / scale[c] / variance[c] + mean[c];
+  }
+}
+
+template <typename T>
+class InplaceHelper {
+ public:
+  template <typename X, typename Scale, typename Bias, typename Mean,
+            typename Variance, typename Y>
+  void operator()(const framework::DataLayout layout, X *x, const Scale *scale,
+                  const Bias *bias, const Mean *mean, const Variance *variance,
+                  double epsilon, int C, int M, const int num, const Y *y,
+                  int grid2, int block, const cudaStream_t &stream) {
+    KeBNRestoreData<<<grid2, block, 0, stream>>>(
+        layout, x, scale, bias, mean, variance, epsilon, C, M, num, y);
+  }
+};
+
+template <>
+class InplaceHelper<paddle::platform::float16> {
+ public:
+  template <typename X, typename Scale, typename Bias, typename Mean,
+            typename Variance, typename Y>
+  void operator()(const framework::DataLayout layout, X *x, const Scale *scale,
+                  const Bias *bias, const Mean *mean, const Variance *variance,
+                  double epsilon, int C, int M, const int num, const Y *y,
+                  int grid2, int block, const cudaStream_t &stream) {
+    PADDLE_THROW("Batch_norm not support in-place for %s for CUDA.",
```

Comment:
Seems that `float16` is totally not supported in batch_norm after this revision. Not a good idea.

Reply:
fixed